### ISSUE TITLE
exe/Main.hs does not work with GHCJS 8.2

### ISSUE THREAD

**Issue by MarcFontaine** [dmjio/miso#414]
The DOM is rendered once and "Hello World !" is printed to the console but no
reaction to click events

**Comment by dmjio** [comment]
I can't reproduce. See here: [Movie.zip](https://github.com/dmjio/miso/files/1934594/Movie.zip)

Which OS / browser are you on?

**Comment by MarcFontaine** [comment]
Thanks for the Movie.

I tried both firefox and google-chrome.
I'm using miso-0.20.0.0  and
x86_64-linux-ghcjs-8.2.0.1 with 
    Cabal-2.0.1.0
    array-0.5.2.0
    base-4.10.1.0
    binary-0.8.5.1
    bytestring-0.10.8.2
    containers-0.5.10.2
    deepseq-1.4.3.0
    directory-1.3.0.2
    filepath-1.4.1.2
    (ghc-8.2.2.20180416)
    ghc-boot-8.2.2.20180416
    ghc-boot-th-8.2.2.20180416
    ghc-prim-0.5.1.1
    ghci-8.2.2.20180416
    ghcjs-prim-0.1.1.0
    ghcjs-th-0.1.0.0
    integer-gmp-1.0.1.0
    pretty-1.1.3.3
    process-1.6.1.0
    rts-1.0
    template-haskell-2.12.0.0
    time-1.8.0.2
    transformers-0.5.2.0
    unix-2.7.2.2

Any Idea where to start debugging ?

**Comment by dmjio** [comment]
Can you send me a screen cap of the behavior? I cannot reproduce on firefox either.

**Comment by dmjio** [comment]
First step, open up the chrome console, see if there are any errors.

**Comment by cocreature** [comment]
GHCJS 8.2 is still quite new so I wouldn’t be surprised if this is a bug.

**Comment by MarcFontaine** [comment]
I'm using ghcjs 8.2 and cabal-install.
It could well be that there is just something wrong with my local setup or that the problem is unrelated to miso.
I'm still new to javascript.
In case you want to look at the problem, I have attached the compiled binaries.
[simple.jsexe.zip](https://github.com/dmjio/miso/files/1935059/simple.jsexe.zip)

The example produced the following output on the firefox console:
* unreachable code after return statement rts.js:7513:2
* The character encoding of the HTML document was not declared. The document will render with garbled text in some browser configurations if the document contains characters from outside the US-ASCII range. The character encoding of the page must be declared in the document or in the transfer protocol.  index.html

* Hello World!  lib.js:8017:9

**Comment by dmjio** [comment]
Event listeners are not being added to `body` for some reason.

**Comment by dmjio** [comment]
@MarcFontaine can you paste your source as well, have a hunch `defaultEvents` isn't being used. Just `"submit"` shows up in list of events, not `"click"`.
<img width="610" alt="screen shot 2018-04-22 at 7 58 10 am" src="https://user-images.githubusercontent.com/875324/39096319-14131b1c-4603-11e8-8fc1-aa89cc367626.png">

**Comment by MarcFontaine** [comment]
This was compiled from the original exe/Main.hs from miso-0.20.0.0.
I did not change anything.
Do you want to full sources of all dependencies?

**Comment by dmjio** [comment]
@MarcFontaine Interesting, yea that would be great. Are you using nix to build this?

**Comment by MarcFontaine** [comment]
At the moment I'm using cabal / cabal sandbox.
I'll try nix next and I'll give you an update if I find out anything new.

**Comment by patrickmn** [comment]
I am experiencing a similar error. GHCJS 8.4.1:

```haskell
defaultEventsXXX :: M.Map MisoString Bool
defaultEventsXXX = M.fromList [
    ("blur", True)
  , ("change", False)
  , ("click", False)
  , ("dblclick", False)
  , ("focus", False)
  , ("input", False)
  , ("keydown", False)
  , ("keypress", False)
  , ("keyup", False)
  , ("mouseup", False)
  , ("mousedown", False)
  , ("mouseenter", False)
  , ("mouseleave", False)
  , ("mouseover", False)
  , ("mouseout", False)
  , ("dragstart", False)
  , ("dragover", False)
  , ("dragend", False)
  , ("dragenter", False)
  , ("dragleave", False)
  , ("drag", False)
  , ("drop", False)
  , ("submit", False)
  ]

foo :: M.Map MisoString String
foo = M.fromList
    [ ("foo", "bar")
    , ("baz", "quux")
    ]

bar :: M.Map String String
bar = M.fromList
    [ ("foo", "bar")
    , ("baz", "quux")
    ]

run :: IO ()
run = do
    print defaultEventsXXX
    print foo
    print bar
    ...
```

outputs

```
fromList [("submit",False)]
fromList [("baz","quux")]
fromList [("baz","quux"),("foo","bar")]
```

something funky about Eq MisoString?

**Comment by patrickmn** [comment]
M.Map Text Text also works fine 😕

**Comment by patrickmn** [comment]
`Ord MisoString` seems messed up:

```haskell
    print $ ("foo" :: MisoString) == ("bar" :: MisoString)
    print $ ("foo" :: MisoString) == ("foo" :: MisoString)
    print $ ("foo" :: MisoString) > ("bar" :: MisoString)
    print $ ("foo" :: MisoString) < ("bar" :: MisoString)
```
->
```
False
True
False
False
```

**Comment by cocreature** [comment]
`MisoString` is just a synonym for `JSString` when you are compiling with `GHCJS`, so I suspect that this is caused by https://github.com/ghcjs/ghcjs/issues/666.

**Comment by FPtje** [comment]
Yeah, it looks like it. It's been fixed in 8.4, so I reckon this issue can be closed.
